src/message.rs

```diff
@@ -0,0 +1,130 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::{BlsScalar, Error, JubJubExtended, JubJubScalar, Note, NoteType};
+
+#[cfg(feature = "canon")]
+use canonical::Canon;
+#[cfg(feature = "canon")]
+use canonical_derive::Canon;
+
+use dusk_jubjub::dhke;
+use dusk_pki::PublicSpendKey;
+use poseidon252::cipher::PoseidonCipher;
+use rand_core::{CryptoRng, RngCore};
+
+/// Message structure with value commitment
+#[derive(Clone, Copy, Debug, Default)]
+#[cfg_attr(feature = "canon", derive(Canon))]
+pub struct Message {
+    value_commitment: JubJubExtended,
+    nonce: JubJubScalar,
+    encrypted_data: PoseidonCipher,
+}
+
+impl Message {
+    /// Create a new message
+    ///
+    /// `r` will be later used to decrypt the value and blinding factor
+    pub fn new<R: RngCore + CryptoRng>(
+        rng: &mut R,
+        r: &JubJubScalar,
+        psk: &PublicSpendKey,
+        value: u64,
+    ) -> Self {
+        let nonce = JubJubScalar::random(rng);
+        let blinding_factor = JubJubScalar::random(rng);
+
+        let note = Note::deterministic(
+            NoteType::Obfuscated,
+            r,
+            nonce,
+            psk,
+            value,
+            blinding_factor,
+        );
+        let Note {
+            value_commitment,
+            nonce,
+            encrypted_data,
+            ..
+        } = note;
+
+        Self {
+            value_commitment,
+            nonce,
+            encrypted_data,
+        }
+    }
+
+    /// Value commitment representation of the message
+    pub fn value_commitment(&self) -> &JubJubExtended {
+        &self.value_commitment
+    }
+
+    /// Nonce used for the encryption of the value and blinding factor
+    pub fn nonce(&self) -> &JubJubScalar {
+        &self.nonce
```

Comment:
As for the other libraries, we avoid exposing the inner attributes of our structures. This approach grant the nonce is read-only since we have no mutation method available.

This, API-wise, means that we can't mutate the nonce. And we indeed can't; if we do, we break the cipher.

Reply:
It isn't consistent with the other structures in this specific library though - for `Note` and `Crossover` most things are `pub(crate)` instead, and the `Fee` has some outright public fields. Even if the field is public, if the structure isn't `mut` then the owner of that structure shouldn't be able to mutate the field either. I'm gonna imagine we should not mutably own this structure once it is created, since all the fields will already be set, so I get what you mean, but I don't think having getters like that is necessary. I will defer the decision to @ZER0 though.

Reply:
After a quick check I think what @vlopes11 is okay: the fields of `Message` struct are not public, if there is no place in the crate where we needs to mutate them (as it happens with `Crossover`) it's okay not having `pub (crate)`. Some structure does not need "getters", however I believe we need them for the `Message` as we have for the `Crossover`.
As first iteration I believe it's okay, there are few things we should change in the API but we can do in the next iteration, when `jubjub` is bumped.

Reply:
Gotcha, makes sense